Allow L2IntegralNorm for subcell

ffoucart opened the pull request:
## Proposed changes

By getting the determinant of the jacobian on the correct grid (subcell or dg), we can take the L2IntegralNorm of a field when using the subcell code.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Upgrade instructions

Any executable that uses ObserverNorm but does not use subcell needs the ComputeTag
` ::Events::Tags::ObserverDetInvJacobianCompute< Frame::ElementLogical, Frame::Inertial>`
in `non_tensor_compute_tags ` 

Any executable that uses ObserverNorm AND subcell needs to have in `non_tensor_compute_tags ` something like
`using non_tensor_compute_tags = tmpl::append< tmpl::conditional_t< use_dg_subcell, tmpl::list< evolution::dg::subcell::Tags::ObserverMeshCompute<volume_dim>, evolution::dg::subcell::Tags::ObserverInverseJacobianCompute< volume_dim, Frame::ElementLogical, Frame::Inertial>, evolution::dg::subcell::Tags::ObserverJacobianAndDetInvJacobian< volume_dim, Frame::ElementLogical, Frame::Inertial>>, tmpl::list<::Events::Tags::ObserverMeshCompute<volume_dim>, ::Events::Tags::ObserverInverseJacobianCompute< volume_dim, Frame::ElementLogical, Frame::Inertial>, ::Events::Tags::ObserverJacobianCompute< volume_dim, Frame::ElementLogical, Frame::Inertial>, ::Events::Tags::ObserverDetInvJacobianCompute< Frame::ElementLogical, Frame::Inertial>>>,`

ffoucart committed:
Allow L2IntegralNorm for subcell

Also add required compute tags to executables using ObserveNorms
and a test of the new FD capabilities of ObserveNorms

nilsdeppe commented:
This looks good. I'm not sure why the tests are failing, though

ffoucart commented:
Some executables need to get the correct event tag apparently.

kidder reviewed (changes requested):
this doesn't work on a subcell grid (it always uses the DetInvJacobian from the DG grid).  You need to add an equivalent `evolution:dg::subcell::Tags` version of the `ObserverDetInvJacobianCompute` that picks the correct thing based on the ActiveGrid.  See Evolution/DgSubcell/Tags/ObserverCoordinates.hpp 
Also this PR needs a unit test

ffoucart commented:
Isn't this already handled by
`evolution::dg::subcell::Tags::ObserverJacobianAndDetInvJacobian<
                  volume_dim, Frame::ElementLogical, Frame::Inertial>`
in Evolution/DgSubcell/Tags/ObserverCoordinates.hpp, together with
`using non_tensor_compute_tags = tmpl::append<
      tmpl::conditional_t<
          use_dg_subcell,
          tmpl::list<
              evolution::dg::subcell::Tags::ObserverMeshCompute<volume_dim>,
              evolution::dg::subcell::Tags::ObserverInverseJacobianCompute<
                  volume_dim, Frame::ElementLogical, Frame::Inertial>,
              evolution::dg::subcell::Tags::ObserverJacobianAndDetInvJacobian<
                  volume_dim, Frame::ElementLogical, Frame::Inertial>>,
          tmpl::list<::Events::Tags::ObserverMeshCompute<volume_dim>,
                     ::Events::Tags::ObserverInverseJacobianCompute<
                         volume_dim, Frame::ElementLogical, Frame::Inertial>,
                     ::Events::Tags::ObserverJacobianCompute<
                         volume_dim, Frame::ElementLogical, Frame::Inertial>,
                     ::Events::Tags::ObserverDetInvJacobianCompute<
                         Frame::ElementLogical, Frame::Inertial>>>,`
in GhValenciaDivCleanBase.hpp (and hopefully other executables using subcell)?

ffoucart commented:
I could certainly add one that _only_ computes the determinant, but as the one computing the determinant and the jacobian is already in the observer box, I'm assuming the jacobian is actually used somewhere.

ffoucart commented:
(I'll look into a unit test, hopefully there is already one for other norms using subcell that I can adapt)

kidder commented:
ah, I didn't look at that tag as it doesn't follow our convention that compute tags have Compute at the end of their name.  Yes that should be fine as long as it is added to the `non_tensor_compute_tags`.  Having the above block of code in either an additional test for L2IntegralNorm or the upgrade instructions of the PR would be helpful

ffoucart commented:
Ok - I added upgrade instruction and tags to all the executables that seem to need them, and I updated the OberverNorm test so that it also uses L2IntegralNorm for finite difference (and the volume as well).

nilsdeppe commented on tests/Unit/ParallelAlgorithms/Events/Test_ObserveNorms.cpp:
```diff
@@ -274,16 +273,23 @@ void test(const std::unique_ptr<ObserveEvent> observe,
     CHECK(results.l2_integral_norm_values[2] == approx(68.22267462752640199));
     CHECK(results.l2_integral_norm_values[3] == approx(95.15951520123110186));
   }
+  else{
```
could you clang-format this block please? The for and else brace spacing specifically, also a space before `{` on line 280

nilsdeppe reviewed (changes requested):
LGTM! Just a formatting request you can squash in immediately
